src/keystore/interfaces.ts

```diff
@@ -0,0 +1,94 @@
+import { messageApi } from '@xmtp/proto'
+import Ciphertext from '../crypto/Ciphertext'
+import {
+  SignedPublicKeyBundle,
+  PublicKeyBundle,
+} from '../crypto/PublicKeyBundle'
+import { InvitationContext } from '../Invitation'
+import { ErrorCode } from './errors'
+
+type KeystoreError = {
+  error: string
+  code: ErrorCode
+}
+
+export type ResultOrError<T> = T | KeystoreError
+
+export type ConversationReference = {
+  topic: string
+  createdAt: Date
+  context?: InvitationContext
+}
+
+export type DecryptV1Request = {
+  payload: Ciphertext
+  peerKeys: PublicKeyBundle
+  headerBytes: Uint8Array
+  isSender: boolean
+}
+
+export type DecryptV1Response = ResultOrError<{
+  decrypted: Uint8Array
+}>
+
+export type DecryptV2Request = {
+  payload: Ciphertext
+  headerBytes: Uint8Array
+  // Need to include contentTopic for the Keystore to know what topic key to use
+  contentTopic: string
+}
+
+export type DecryptV2Response = ResultOrError<{
+  decrypted: Uint8Array
+}>
+
+export type EncryptV1Request = {
+  recipient: PublicKeyBundle
+  payload: Uint8Array
+  headerBytes: Uint8Array
+}
+
+export type EncryptResponse = ResultOrError<{
+  ciphertext: Ciphertext
+}>
+
+export type EncryptV2Request = {
+  contentTopic: string
+  message: Uint8Array
+  headerBytes: Uint8Array
+}
+
+export type CreateInviteRequest = {
+  recipient: SignedPublicKeyBundle
+  createdAt: Date
+  context: InvitationContext
+}
+
+export type CreateInviteResponse = {
+  conversation: ConversationReference
+  // The full bytes of the sealed invitation, which can then be published to the API
+  payload: Uint8Array
+}
+
+export interface Keystore {
+  // Decrypt a batch of V1 messages
+  decryptV1(req: DecryptV1Request[]): Promise<DecryptV1Response[]>
+  // Decrypt a batch of V2 messages
+  decryptV2(req: DecryptV2Request[]): Promise<DecryptV2Response[]>
+  // Encrypt a batch of V1 messages
+  encryptV1(req: EncryptV1Request[]): Promise<EncryptResponse[]>
+  // Encrypt a batch of V2 messages
+  encryptV2(req: EncryptV2Request[]): Promise<EncryptResponse[]>
+  // Decrypt and save a batch of invite for later use in decrypting messages on the invite topic
+  saveInvites(
+    req: messageApi.Envelope[]
+  ): Promise<ResultOrError<ConversationReference>[]>
+  // Create the sealed invite and store the Topic keys in the Keystore for later use
+  createInvite(req: CreateInviteRequest): Promise<CreateInviteResponse>
+  // Get V2 conversations
+  getV2Conversations(): Promise<ConversationReference[]>
+  // Used for publishing the contact
+  getPublicKeyBundle(): Promise<SignedPublicKeyBundle>
```

Comment:
So, I am assuming that all Keystore implementations will have access to a `PrivateKeyBundle`. That means they will always have a `PublicKeyBundle` they can return

Reply:
Great question. I am assuming that each Keystore implementation would be responsible for bootstrapping itself with the `PrivateKeyBundle` (which is the XMTP private keys). 

The `InMemoryKeystore` would perform that bootstrapping by downloading the `EncryptedPrivateKeyBundle` from the network and decrypting it with a wallet signature (the same as we do right now). Other Keystore implementations might only have to do that once, and could then start up subsequent times by loading the keys from secure storage (for example, the MM Snap)

Reply:
Generally I see bootstrapping as something that lives outside the Keystore API. It will be different between implementations, and should be designed as a process that doesn't _require_ the SDK to touch any secrets.